language/resource-api/apt_key.rb

```diff
@@ -0,0 +1,375 @@
+            docs:    'The key server to fetch the key from based on the ID. It can either be a domain name or url.',
+            default: 'keyserver.ubuntu.com'
+        },
+        options:     {
+            type: 'Optional[String]',
+            docs: 'Additional options to pass to apt-key\'s --keyserver-options.',
+        },
+        fingerprint: {
+            type:      'String',
+            docs:      'The 40-digit hexadecimal fingerprint of the specified GPG key.',
+            read_only: true,
+        },
+        long:        {
+            type:      'String',
+            docs:      'The 16-digit hexadecimal id of the specified GPG key.',
+            read_only: true,
+        },
+        short:       {
+            type:      'String',
+            docs:      'The 8-digit hexadecimal id of the specified GPG key.',
+            read_only: true,
+        },
+        expired:     {
+            type:      'Boolean',
+            docs:      'Indicates if the key has expired.',
+            read_only: true,
+        },
+        expiry:      {
+            # TODO: should be DateTime
+            type:      'String',
+            docs:      'The date the key will expire, or nil if it has no expiry date, in ISO format.',
+            read_only: true,
+        },
+        size:        {
+            type:      'Integer',
+            docs:      'The key size, usually a multiple of 1024.',
+            read_only: true,
+        },
+        type:        {
+            type:      'String',
+            docs:      'The key type, one of: rsa, dsa, ecc, ecdsa.',
+            read_only: true,
+        },
+        created:     {
+            type:      'String',
+            docs:      'Date the key was created, in ISO format.',
+            read_only: true,
+        },
+    },
+    autorequires: {
+        file:    '$source', # will evaluate to the value of the `source` attribute
+        package: 'apt',
+    },
+}
+
+module Puppet::SimpleResource
+  class TypeShim
+    attr_reader :values
+
+    def initialize(title, resource_hash)
+      # internalize and protect - needs to go deeper
+      @values        = resource_hash.dup
+      # "name" is a privileged key
+      @values[:name] = title
+      @values.freeze
+    end
+
+    def to_resource
+      ResourceShim.new(@values)
+    end
+
+    def name
+      values[:name]
+    end
+  end
+
+  class ResourceShim
+    attr_reader :values
+
+    def initialize(resource_hash)
+      @values = resource_hash.dup.freeze # whatevs
+    end
+
+    def title
+      values[:name]
+    end
+
+    def prune_parameters(*args)
+      puts "not pruning #{args.inspect}" if args.length > 0
+      self
+    end
+
+    def to_manifest
+      [
+          "apt_key { #{values[:name].inspect}: ",
+      ] + values.keys.select { |k| k != :name }.collect { |k| "  #{k} => #{values[k].inspect}," } + ['}']
+    end
```

Comment:
this feels very clear and logical, but the rest up there just reads like boilerplate…
what am i missing?